google-cloud-firestore/src/test-jdk17/java/com/google/cloud/firestore/RecordDocumentReferenceTest.java

```diff
@@ -0,0 +1,407 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import static com.google.cloud.firestore.LocalFirestoreHelper.assertCommitEquals;
+import static com.google.cloud.firestore.LocalFirestoreHelper.commit;
+import static com.google.cloud.firestore.LocalFirestoreHelper.create;
+import static com.google.cloud.firestore.LocalFirestoreHelper.getAllResponse;
+import static com.google.cloud.firestore.LocalFirestoreHelper.map;
+import static com.google.cloud.firestore.LocalFirestoreHelper.serverTimestamp;
+import static com.google.cloud.firestore.LocalFirestoreHelper.set;
+import static com.google.cloud.firestore.LocalFirestoreHelper.transform;
+import static com.google.cloud.firestore.RecordTestHelper.*;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
+
+import com.google.api.gax.rpc.ResponseObserver;
+import com.google.api.gax.rpc.ServerStreamingCallable;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.cloud.firestore.spi.v1.FirestoreRpc;
+import com.google.firestore.v1.BatchGetDocumentsRequest;
+import com.google.firestore.v1.CommitRequest;
+import com.google.firestore.v1.CommitResponse;
+import com.google.firestore.v1.Value;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.ArgumentMatchers;
+import org.mockito.Captor;
+import org.mockito.Mockito;
+import org.mockito.Spy;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class RecordDocumentReferenceTest {
+
+  @Spy
+  private final FirestoreImpl firestoreMock =
+      new FirestoreImpl(
+          FirestoreOptions.newBuilder().setProjectId("test-project").build(),
+          Mockito.mock(FirestoreRpc.class));
+
+  @Captor private ArgumentCaptor<CommitRequest> commitCapture;
+
+  @Captor private ArgumentCaptor<BatchGetDocumentsRequest> getAllCapture;
+
+  @Captor private ArgumentCaptor<ResponseObserver<CommitResponse>> streamObserverCapture;
+
+  private DocumentReference documentReference;
+
+  @Before
+  public void before() {
+    documentReference = firestoreMock.document("coll/doc");
+  }
+
+  @Test
+  public void serializeBasicTypes() throws Exception {
+    doReturn(SINGLE_WRITE_COMMIT_RESPONSE)
+        .when(firestoreMock)
+        .sendRequest(
+            commitCapture.capture(), ArgumentMatchers.<UnaryCallable<CommitRequest, CommitResponse>>any());
+
+    documentReference.set(ALL_SUPPORTED_TYPES_OBJECT).get();
+
+    CommitRequest expectedCommit = commit(set(ALL_SUPPORTED_TYPES_PROTO));
```

Comment:
Since these files are only compile-able with JDK16+ anyway, why not use `var`?

Reply:
Explicitly writing out the types is more readable and maintainable. Clarity is preferred over simplicity.